### PR TITLE
Deprecate index merging

### PR DESCRIPTION
**Pull request by vyasr**
This PR deprecates support for merging Index objects. pandas only supports merging of DataFrames, so we should move towards that as well. The main internal implication of this change is that `BaseIndex.union` and `BaseIndex.difference` now require an internal conversion to a `DataFrame` followed by a conversion of the result back to the appropriate index type. Since the intermediate objects are not modified and don't involve additional memory allocations, this change just adds a little bit of Python overhead to index merging (10-50 us). Once the deprecated code is fully removed, though, we should be able to make this time back by simplifying the internals of joining, which currently has logic for handling Series and Index objects internally.

### CODE REVIEW

File: python/cudf/cudf/core/_base_index.py [hunk 1]
```diff
@@ -707,7 +708,18 @@ def difference(self, other, sort=None):
...
         else:
-            difference = self.join(other, how="leftanti")
+            other = other.copy(deep=False)
+            other.names = self.names
+            difference = cudf.core.index._index_from_data(
+                cudf.DataFrame._from_data(self._data)
+                ._merge(
+                    cudf.DataFrame._from_data(other._data),
+                    how="leftanti",
+                    on=self.name,
```
Comment (isVoid) [inline python/cudf/cudf/core/_base_index.py:718]: Should this be `self.names`? My understanding is that to make this method also work for multiindex you need to merge on all columns.

Comment (vyasr) [reply python/cudf/cudf/core/_base_index.py:718]: You're right, but I'm pretty sure MultiIndex-MultiIndex joining is already broken on our main branch. MultiIndex-Index merging happens to work, and we have a test for that, so I made sure not to break that, but I didn't want to expend effort making MultiIndex-MultiIndex merging work if it's a) already broken and b) about to be removed in the next release anyway.

Comment (isVoid) [reply python/cudf/cudf/core/_base_index.py:718]: That sounds reasonable to me.

**Review by isVoid — APPROVED** [review]

**Comment by vyasr** [comment]
@gpucibot merge

**Pull request merged**